### PR TITLE
Use the host path when current_folder points to the document storage 

### PR DESCRIPTION
**Pull request by xhorak**
Addressing issue https://github.com/flatpak/xdg-desktop-portal/issues/982

### CODE REVIEW

**Comment by ebassi** [comment]
Please, conform to the coding style of the rest of the project.

File: src/file-chooser.c [hunk 1]
```diff
@@ -665,6 +664,50 @@ handle_save_file (XdpDbusFileChooser *object,
...
+                while (g_variant_iter_loop(doc_iter, "{s@ay}", &doc_id,
+                                           &doc_path)) {
+                  if (g_strcmp0(doc_id_from_app, doc_id) == 0) {
+                    // Chop off the filename from the host path.
+                    const char *doc_path_str =
+                        g_variant_get_bytestring(doc_path);
+
+                    g_free(host_path);
+                    char *filename = strrchr(doc_path_str, '/');
+                    if (filename) {
```
Comment (ebassi) [inline src/file-chooser.c:693]:
```suggestion
                    if (filename)
                      {
```

File: src/file-chooser.c [hunk 2]
```diff
@@ -665,6 +664,50 @@ handle_save_file (XdpDbusFileChooser *object,
...
+                      host_path =
+                          g_strndup(doc_path_str, filename - doc_path_str);
```
Comment (ebassi) [inline src/file-chooser.c:695]:
```suggestion
                          g_strndup (doc_path_str, filename - doc_path_str);
```

Comment (ebassi) [reply src/file-chooser.c:695]: This is just `g_path_get_dirname()` with extra steps.

File: src/file-chooser.c [hunk 3]
```diff
@@ -665,6 +664,50 @@ handle_save_file (XdpDbusFileChooser *object,
...
+                    } else {
+                      host_path = g_strdup(doc_path_str);
+                    }
+                    break;
+                  }
+                }
+                g_variant_iter_free(doc_iter);
```
Comment (ebassi) [inline src/file-chooser.c:702]:
```suggestion
                g_variant_iter_free (doc_iter);
```

File: src/file-chooser.c [hunk 4]
```diff
@@ -665,6 +664,50 @@ handle_save_file (XdpDbusFileChooser *object,
...
+                g_debug("SaveFile: translating current_folder value '%s' to host path '%s'",
```
Comment (ebassi) [inline src/file-chooser.c:703]:
```suggestion
                g_debug ("SaveFile: translating current_folder value '%s' to host path '%s'",
```

File: src/file-chooser.c [hunk 5]
```diff
@@ -665,6 +664,50 @@ handle_save_file (XdpDbusFileChooser *object,
...
+                    path_from_app, host_path);
+              }
+        }
+        g_variant_builder_add(&options, "{sv}", "current_folder",
```
Comment (ebassi) [inline src/file-chooser.c:707]:
```suggestion
        g_variant_builder_add (&options, "{sv}", "current_folder",
```

File: src/file-chooser.c [hunk 6]
```diff
@@ -665,6 +664,50 @@ handle_save_file (XdpDbusFileChooser *object,
...
+                              g_variant_new_bytestring(host_path));
```
Comment (ebassi) [inline src/file-chooser.c:708]:
```suggestion
                              g_variant_new_bytestring (host_path));
```

**Review by ebassi — CHANGES_REQUESTED** [review]
Would have been faster, if github had the ability to deal with suggestions involving more than one line.

**Comment by xhorak** [comment]
Addressed also: https://github.com/xhorak/xdg-desktop-portal/commit/b3c3f5610ea68b0a13b371d1dff16403d0f8f7e6#r111467974

File: src/file-chooser.c [hunk 7]
```diff
@@ -601,6 +600,24 @@ looks_like_document_portal_path (const char *path,
   return TRUE;
 }
 
+static char*
+get_host_folder_for_doc_id (const char* doc_id) {
```
Comment (matthiasclasen) [inline src/file-chooser.c:604]:
```suggestion
get_host_folder_for_doc_id (const char* doc_id)
{
```

File: src/file-chooser.c [hunk 8]
```diff
@@ -601,6 +600,24 @@ looks_like_document_portal_path (const char *path,
...
+  gchar *host_folder = NULL;
+  g_autofree char *real_path = get_real_path_for_doc_id (doc_id);
+  if (real_path) {
```
Comment (matthiasclasen) [inline src/file-chooser.c:607]:
```suggestion
  if (real_path)
    {
```

File: src/file-chooser.c [hunk 9]
```diff
@@ -601,6 +600,24 @@ looks_like_document_portal_path (const char *path,
...
+    char *filename = strrchr (real_path, '/');
```
Comment (matthiasclasen) [inline src/file-chooser.c:608]:
```suggestion
      char *filename = strrchr (real_path, '/');
```

File: src/file-chooser.c [hunk 10]
```diff
@@ -665,6 +680,28 @@ handle_save_file (XdpDbusFileChooser *object,
         g_variant_builder_add (&options, "{sv}", "current_file", g_variant_new_bytestring (host_path));
       }
   }
+  {
+      g_autoptr(GVariant) value =
+        g_variant_lookup_value (arg_options, "current_folder", G_VARIANT_TYPE_BYTESTRING);
+
+      if (value) {
```
Comment (matthiasclasen) [inline src/file-chooser.c:687]: Niggle: move the { to the next line

Comment (matthiasclasen) [reply src/file-chooser.c:687]: Apart from that, this looks good to me now.

Comment (jadahl) [reply src/file-chooser.c:687]: The indentation is wrong too

**Comment by jadahl** [comment]
The commits should be squashed and turned into the first one before landing.

File: src/file-chooser.c [hunk 11]
```diff
@@ -601,6 +600,22 @@ looks_like_document_portal_path (const char *path,
...
+{
+  gchar *host_folder = NULL;
+  g_autofree char *real_path = get_real_path_for_doc_id (doc_id);
+  if (real_path)
+    {
+      char *filename = strrchr (real_path, '/');
+      if (filename)
+        host_folder = g_strndup (real_path, filename - real_path);
+      else
+        host_folder = g_strdup (real_path);
```
Comment (ebassi) [inline src/file-chooser.c:614]: This is `g_path_get_dirname()`

**Comment by matthiasclasen** [comment]
From my perspective, this is good to go

**Comment by ebassi** [comment]
Same here, so I guess I get to press the green button. :smile:

**Comment by lissyx** [comment]
So while it helps re-open the folder with a "nice name" we still save an XDG document path. Is it possible we have this feature exposed somehow ? Via DBus ? Or maybe it is already doable ? @matthiasclasen @ebassi ?

**Comment by ssokolow** [comment]
Isn't unprivileged code always seeing an XDG document path part of the point for ensuring that unnecessary information that may be contained in the names of ancestor folders isn't leaked?

**Comment by ebassi** [comment]
@lissyx Please, open an issue. Commenting on a closed merge request isn't going to help in tracking issues.

**Comment by lissyx** [comment]
> @lissyx Please, open an issue. Commenting on a closed merge request isn't going to help in tracking issues.

I was asking a question, not reporting a bug

**Pull request merged**